### DISCUSSION TITLE
[nuxt3] .env or publicRuntime inside composables?

### DISCUSSION THREAD

**Discussion by ctwhome** [nuxt/nuxt#17050] [Questions]
is there a way to pass .env variables to a composable?
I want to simply use Supabase within a composable and I am trying different ways but I can't get it right.

repo: https://github.com/ctwhome/top-nuxt3

`composables/useSupabase`:
```js
import { createClient } from '@supabase/supabase-js'

const SUPABASE_URL = 'https://___.supabase.co'      <-- REPLACE STRING with .env
const SUPABASE_KEY = 'ey___'                        <-- REPLACE STRING with .env

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY)

export default function useSupabase () {
  return { supabase }
}
```
I just want to use the const supabase inside this composable 
`composables/useAuth`:
```js
import useSupabase from '~/composables/useSupabase'

export default function useAuth() {
  const {supabase} = useSupabase()
...
```

I also try to use it inside a plugin but I don't know how to call it inside the composable.

`plugins/supabase.client.ts`:
```js
import {defineNuxtPlugin} from '#app'
import { createClient } from '@supabase/supabase-js'

export default defineNuxtPlugin(nuxtApp => {
  const config = useRuntimeConfig()
  const supabase = createClient(config.supabaseUrl, config.supabasePublicKey)
  nuxtApp.provide('supabase',supabase)
  return supabase
})

```

**Comment by ctwhome** [comment]
@danielroe would you have a tip for this? 👐

**Reply by GoudekettingRM** [reply]
Did you find out how to get this to work? Is there a solution for this already?

**Reply by ctwhome** [reply]
I went for sveltekit 😬

**Reply by GoudekettingRM** [reply]
Hahaha, alright, so that's a no 😂  Thanks for the reply though!

**Comment by danielroe** [comment]
Something like this should probably work - just POC - don't call `useRuntimeConfig` twice, for example!

`composables/useSupabase`:
```diff
  import { createClient } from '@supabase/supabase-js'
+ import { useRuntimeConfig } from '#app'

- const SUPABASE_URL = 'https://___.supabase.co'      <-- REPLACE STRING with .env
- const SUPABASE_KEY = 'ey___'                        <-- REPLACE STRING with .env

+ let supabase
-  const supabase = createClient(SUPABASE_URL, SUPABASE_KEY)

  export default function useSupabase () {
+   supabase = supabase || createClient(useRuntimeConfig().SUPABASE_URL, useRuntimeConfig().SUPABASE_KEY)
    return { supabase }
  }
```

**Reply by ctwhome** [reply]
thanks for the reply! 

It looks like it is not possible to call useRuntimeConfig(), It imports the function, but if I call it with the parenthesis at the end it gives a 500 error: 
```
nuxt instance unavailable
  at useNuxtApp (file://./.nuxt/dist/server/server.mjs:315:13)  
  at Module.useRuntimeConfig (file://./.nuxt/dist/server/server.mjs:323:10)  
  at $id_c8de8b8d (file://./.nuxt/dist/server/server.mjs:2704:67)  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:16696:3)
```

I made a branch for it here: https://github.com/ctwhome/top-nuxt3/tree/env-composables

**Reply by danielroe** [reply]
You should be using it within a Nuxt plugin or setup function.

**Reply by ctwhome** [reply]
I am afraid I can't get it right. on nuxt 2 I `const export supabase` from my plugin and import it outside. 

However, I don't know how to use the plugin  inside the composable to use the supabase variable:
`plugins/supabase.ts`

```js
import {defineNuxtPlugin, useRuntimeConfig} from '#app'
import { createClient } from '@supabase/supabase-js'

export default defineNuxtPlugin(nuxtApp => {
  const supabase = createClient( useRuntimeConfig().SUPABASE_URL, useRuntimeConfig().SUPABASE_PUBLIC_KEY )
  return { supabase }
})

```
`composables/useAuth`:
```js
---- option 1
import { useNuxtApp } from '#app'
const { supabase } = useNuxtApp()   --> Doesn't' work

---- option 2
import {supabase} from '~/plugins/supabase'  --> this doesn't work either

```

**Reply by Imanullah** [reply]
`composables/useSupabase.ts`

export default function () {
  const nuxtApp = useNuxtApp()
  return nuxtApp.$supabase
}

on my personal problem is how to use composable or plugin in 
Store folder files for Pinia

**Reply by ctwhome** [reply]
@Imanullah where do you call exactly the env vars to create the client for supabase inside the composable?

**Reply by mdhelaluddin-ctg-bd** [reply]
@ctwhome any update on the issue? Did you able to solve this?

**Reply by ctwhome** [reply]
no, I didn't

**Reply by LuckyWindsck** [reply]
> `composables/useSupabase.ts`
> 
> ```TypeScript
> export default function () {
>   const nuxtApp = useNuxtApp()
>   return nuxtApp.$supabase
> }
> ```
> 
> on my personal problem is how to use composable or plugin in Store folder files for Pinia

I have same problem as yours, and I found this way solve my problem.

It seems like that runtime config can not be access in composable files. So I first create a plugin, and call `useRuntimeConfig()` inside `defineNuxtPlugin(...)`. In this way, you can get the `variable` you provide in the plugin with `$variable` from `useNuxtApp()`.

So in your case, I guess the following code should work:

`plugins/supabase.client.ts`:
```TypeScript
import { createClient } from '@supabase/supabase-js'

export default defineNuxtPlugin(() => {
  const { supabaseUrl, supabasePublicKey } = useRuntimeConfig()
  const supabase = createClient(supabaseUrl, supabasePublicKey)

  return { provide: { supabase } }
})
```

`composables/useAuth.ts`:
```TypeScript
export default useAuth = () => {
  const { $supabase } = useNuxtApp()

  // Do anything you want with the $supabase client
}
```

**Comment by LuckyWindsck** [comment]
Maybe we just need to provide `useRuntimeConfig` to NuxtApp, then we can call `useNuxtApp().$getRuntimeConfigInComposable()` to get runtime config. I am not familiar with Nuxt3 lifecycle yet, but somehow it works for me.

`plugins/getRuntimeConfigInComposable.ts`

```TypeScript
export default defineNuxtPlugin(() => ({
  provide: {
    getRuntimeConfigInComposable: useRuntimeConfig,
  },
}))
```

`composables/anyComposable.ts`
```TypeScript
export default () => {
  const { $getRuntimeConfigInComposable } = useNuxtApp()
  const config = $getRuntimeConfigInComposable()

  // Use `config` to do whatever thing you want.
}
```

**Reply by GoudekettingRM** [reply]
Yeah, that might work, but my gosh what an inelegant solution. Nuxt3 enables auto imports of most things into most other things, but somehow `process.env` or `import.meta.env` is not available in a new top level folder? Kind of rubs me the wrong way 😅

Hopefully on first release that might be included?

**Reply by danielroe** [reply]
See https://github.com/nuxt/framework/discussions/3215#discussioncomment-3082845.

**Comment by danielroe** [comment]
Anywhere you can can call `useNuxtApp()` you can also call `useRuntimeConfig()`, and you can also call it ambiently within server routes.

There are good reasons that we don't just proxy all your `process.env` into your app. (Everything included would have to be passed from server to client in the HTML payload - and much of that may be sensitive information, for example.)

For more info, check out https://v3.nuxtjs.org/guide/features/runtime-config/.

**Reply by GoudekettingRM** [reply]
I thought I tried just calling `useRuntimeConfig()`, but I must've mucked something up.. 

Thanks for the reply in any case, I understand that there are a lot of considerations. And with that being said, I do like what the rc is so far!

**Reply by LuckyWindsck** [reply]
Thank you for your explanation.

Also, I found the reason why my calling of `useRuntimeConfig()` in composable file failed. I should call it inside the composable function. So it seems like the contexts inside and outside of the composable function are different.

Wrong way:
```Typescript
const config = useRuntimeConfig()

export default () => {
  console.log(config)
}
```

Correct way:
```Typescript
export default () => {
  const config = useRuntimeConfig()

  console.log(config)
}
```

**Comment by timothybennette8** [comment]
Did anyone find a solution to accessing .env directly within a Composable?
I've been through all of the above with no success - creating a plugin and calling `useRuntimeConfig()` within a Composable function both return `no nuxt instance available`.
The best I've been able to achieve is passing them from the page/component but obviously this is far from ideal!

**Reply by danielroe** [reply]
The issue is likely that you are doing that in an async function. Composables, with a few exceptions, should be synchronous. And if not synchronous, you should always get runtime config _before_ awaiting a promise.

**Reply by timothybennette8** [reply]
I was getting a bit lost and so I went for a very simple approach to see what works:

A getter function for env variables
`composables/useEnv.ts`
```javascript
export default function () {
  const nuxtApp = useNuxtApp();

  return {
    EXAMPLE_VARIABLE: nuxtApp.$config.public.EXAMPLE_VARIABLE,
    SUPABASE_KEY: nuxtApp.$config.public.SUPABASE_KEY,
    SUPABASE_URL: nuxtApp.$config.public.SUPABASE_URL,
    SUPABASE_SERVICE_ROLE_KEY: nuxtApp.$config.public.SUPABASE_SERVICE_ROLE_KEY,
  };
}
```

A composable to call the getter function and return the result from both a synchronous and an asynchronous function
`composables/anotherComposable.ts`
```javascript
export default function () {
  const synchronous = () => {
    const { EXAMPLE_VARIABLE } = useEnv();

    return {
      EXAMPLE_VARIABLE,
    };
  };

  const asynchronous = async () => {
    const { EXAMPLE_VARIABLE } = useEnv();

    return {
      EXAMPLE_VARIABLE,
    };
  };
  return {
    synchronous,
    asynchronous,
  };
}
```

A composable using an async function to call the previous composable's sync and async functions and return
`composables/furtherComposable.ts`
```javascript
export default function () {
  const asyncTestFunc = async () => {
    const { synchronous, asynchronous } = anotherComposable();

    return {
      synchronousResult: synchronous(),
      asynchronousResult: await asynchronous(),
    };
  };

  return {
    asyncTestFunc,
  };
}
```
Call the results of all functions in a page
`pages/index.vue`
```html
<script setup lang='ts'>
const { EXAMPLE_VARIABLE } = useEnv();

const { synchronous, asynchronous } = anotherComposable();

const { asyncTestFunc } = furtherComposable();

console.log("1", EXAMPLE_VARIABLE)

console.log("2", synchronous())

console.log("3", await asyncTestFunc())
</script>
```
![example_variable](https://user-images.githubusercontent.com/43624464/184509864-c56af3d7-0715-4991-8259-339b2ee6aac4.jpg)

So far, so good 😃 

But the following returns `nuxt instance unavailable` - I don't understand why `asyncTestFunction()` works but this doesn't. 😞 

A composable to create a Supabase client
`composables/useSupabase.ts`
```javascript
import { createClient } from "@supabase/supabase-js";

export default function () {
  const { SUPABASE_KEY, SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY } = useEnv();
 
  const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
  const supabaseAdmin = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY);
  return {
    supabase,
    supabaseAdmin,
  };
}
```

A composable to perform CRUD operations using Supabase client
`composables/useCRUD.ts`
```javascript
const { supabase } = useSupabase();

const useCRUD = () => {
  const getItems = async (table) => {
    const res = await supabase.from(table).select();
    return res;
  };
  return {
    getItems
  };
};

export default useCRUD;
```

**Reply by timothybennette8** [reply]
After more experimentation I've got to a point where it "works" but I don't think I've taken the right approach at all 🙃 

I have a supabase client plugin that gets my env variables and returns both instance of a client:
`plugins/supabase.client.ts`
```javascript
import { createClient } from "@supabase/supabase-js";

export default defineNuxtPlugin(() => {
  const { SUPABASE_URL, SUPABASE_KEY, SUPABASE_SERVICE_ROLE_KEY } =
    useRuntimeConfig();

  const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
  const supabaseAdmin = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY);

  return { provide: { supabase, supabaseAdmin } };
});
```

Then I have a simple composable to handle supabase storage getters (note that this gets an additional env variable):
`composables/useStorage`
```javascript
const useStorage = () => {
  const { $supabase } = useNuxtApp();
  const { SUPABASE_BUCKET: bucket } = useRuntimeConfig();

  const getPublicURL = async (path, filename) => {
    const { publicURL, error } = await $supabase.storage
      .from(bucket)
      .getPublicUrl(`${path}/${filename}`);
    return publicURL;
  };

  const getFiles = async (path) => {
    const { $supabase } = useNuxtApp();

    const { data, error } = await $supabase.storage.from(bucket).list(path, {
      limit: 100,
      offset: 0,
      sortBy: { column: "name", order: "asc" },
    });

    return data;
  };

  return {
    getPublicURL,
    getFiles,
    getImages,
  };
};

export default useStorage;

```

Then in a page I call the composable:
`pages/index.vue`
```html
<script setup>
const { getPublicURL, getFiles } = useStorage();

const url = ref("")
getPublicURL("<PATH TO FILE>", "<FILENAME>").then(res => url.value = res)

const images = ref([])
images.value = getImages("<PATH>").then(res => images.value = res)

</script>

<template>
  <div class="container">
      <img :src="url">
      <img v-for="img in images" :src="img" />
   </div>
</template>
```

The only way I can get the returned values to be accessible in the page is to assign them to a ref() via a then() method - root level await doesn't work. Judging by the two errors I get in VS Code terminal the client isn't available in my composables at some point in the lifecyle:
```bash
[nitro] [dev] [unhandledRejection] TypeError: Cannot read property 'storage' of undefined
    at getPublicURL
    // stack trace removed for brevity
    at getFiles
```
I get that this is _probably_ because the client isn't returning during the initial page render, but assigning the then() methods to ref()s allows them to be populated later on (after page load?).
Obviously, the errors indicate that I've not got this quite right despite being able to render the API data in the browser successfully.
I've also tried using useAsyncData() to call the API within synchronous functions but that doesn't return anything. 
I can't figure out how I can wrap it all up in composables if I can't await the response at page level. 
Any guidance would be greatly appreciated at this point as I have very much hit the proverbial brick wall!
